### ISSUE TITLE
SSR production mode is slower than SSR development mode

### ISSUE THREAD

**Issue by balciseri** [reactql/cli#47]
Hello, today i've spent quite a bit of time reading about SSR performances because I've noticed that running my project with "npm start" and visiting the SSR page (which is now in dev mode ) versus running "npm run build-run" and visiting the SSR page ( which is now in prod mode) gave weird results.
In quite a lot of guides they say that just building react from dev mode to prod mode should give you a 300-400% boost in perfs...
But the more i tried to refresh the page the more for me it felt slower in prod mode.
So i added a simple code to test `ReactDOMServer.renderToString(components);`
```
 // Full React HTML render
    const start2 = ms.now();
    const html = ReactDOMServer.renderToString(components);
    const end2 = ms.parse(ms.since(start2));
    const total2 = end2.microseconds + (end2.milliseconds * 1e3) + (end2.seconds * 1e6);
    console.log(`${total2 / 1e3}ms`);
```
At first  I tought that I've must done some errors somewhere.. so today i've updated reactql and started a fresh project and added the bare code needed ro reproduce the issue: in my case i'm rendering a tree data structure with `<ul>`s and `<li>`s. The more nodes  the tree had the slower the production mode would be in executing `const html = ReactDOMServer.renderToString(components);`.

TL;DR: Why reactql in Production is slower to run `ReactDOMServer.renderToString(components);` than the Dev mode the more components you have to render?

I know this will be hard to read, i'm here if you need me to explain or provide more infos

Edit: apart for the console.log to measure the ms of the function this is the only code i've added:
http://puu.sh/wwFGr/dcb5f6e74d.png

This is issue is quite sersious because in the example i only used stateless components, but this is even more problematic when i need to have also state inside them to handle toggles, collapse, checkboxes, calculations ecc

Edit2: the data i've got 
http://puu.sh/wwGey/e424a68cd2.png

And in my real project where the nodes are not this simple is way more problematic. How is this possible that in Prod mode it will take twice the time ?

**Comment by leebenson** [comment]
interesting!

I can't replicate the issue locally. Using the [graphql-server example](https://github.com/reactql/examples/tree/master/graphql-server) (which runs a local GraphQL server locally to avoid external calls), and using this code:

```js
    const start = ms.now();

    const html = ReactDOMServer.renderToString(components);

    const end = ms.parse(ms.since(start));
    const total = end.microseconds + (end.milliseconds * 1e3) + (end.seconds * 1e6);

    console.log('total ->', total);
```

... I get:

> DEV:
> 
> total -> 18814
> total -> 10561
> total -> 5495
> total -> 4157
> total -> 9165
> total -> 7760
> total -> 5350
> total -> 4340
> total -> 5099
> 
> PROD:
> 
> total -> 21876
> total -> 5266
> total -> 4608
> total -> 5091
> total -> 5705
> total -> 4123
> total -> 4535
> total -> 2980
> total -> 5559

The first call in each call is an outlier- the code is running 'cold' for the first time. Thereafter, once the VM has warmed up, requests are very comparable in both scenarios.

Using a fresh project (kit v1.10.1) and the default graph.cool endpoint, I get:

> DEV:
> 
> total -> 19687
> total -> 8904
> total -> 5892
> total -> 10254
> total -> 6899
> total -> 10005
> total -> 10774
> total -> 6983
> total -> 7133
> 
> PROD:
> 
> total -> 15903
> total -> 7994
> total -> 6186
> total -> 7194
> total -> 4773
> total -> 6585
> total -> 6478
> total -> 6564
> total -> 6011

Again, too close to call - 5-7ms is quite common.

If you're timing solely the React render chain _after_ the GraphQL has already been fetched and injected (as I'm doing here), then it could be an issue that exists in Apollo or React land-- there's no ReactQL code that would slow either environment down.

**Comment by balciseri** [comment]
Man I love your effort and activity on this project. 
Your dedication is amazing and you are quite brilliant.

Sorry if i've gave the blame to ReactQl which of course is not the culprit here.

I did the same test again as you did. "reactql new" added your monitoring code and even as minimal i still see that `renderToString` in production mode is slower (this should be totally impossible). This is so weird since everywhere it's stated that react in Dev mode spends 3/4 of his time just doing error checking. So Prod mode beeing slower is totally wrong. React is doing something wrong. And i'm so sad about this because i don't know how to address the fact that my app ( ANY app) is running slower in Prod mode which should not, actualyl it should be the opposite.

Ofcourse with the empty new reactql app you can't visually feel that is running slower but as soon as you start to build a complex page ( Like mine was) where sometimes you need to render a not so small ( where virtualization it's still not needed) tree of data the time almost doubles.

Basically the more work `renderToString` has to do the slower the Production mode will be and this makes me so downhearted.

I'll add the same test you did. React should be faster in Prod mode no matter what , not the opposite. "React why? what are you doing ??"
```
webpack: Compiled successfully.
Dev Mode
total -> 20466
total -> 9813
total -> 7609
total -> 9010
total -> 5323
total -> 6876
total -> 7859
total -> 6235
total -> 10106
total -> 6184
total -> 5925
total -> 3598
total -> 4975
total -> 4162
total -> 3602
total -> 6384
total -> 3519
total -> 3770
total -> 9007
total -> 3542
total -> 3334
total -> 7639
total -> 4865
total -> 3668
total -> 5287
total -> 3455
total -> 4428
total -> 4756
total -> 3534
```
```
Prod mode
total -> 22465
total -> 9640
total -> 8038
total -> 9803
total -> 9051
total -> 8708
total -> 8444
total -> 7550
total -> 7440
total -> 7410
total -> 6837
total -> 6561
total -> 5424
total -> 5596
total -> 5655
total -> 5238
total -> 7040
total -> 6536
total -> 9661
total -> 5420
total -> 5357
total -> 6953
total -> 5157
total -> 6538
```
Thanks again for your awesome work, I know this is not on you. But this is a serious problem and i'm not able to take it forward to the right people.

**Comment by leebenson** [comment]
I'll re-open this to encourage external feedback; this isn't something I'm able to replicate locally using node 8.1.2 on macOS Sierra 10.12.6 beta.

Which version of Node are you using? I'm wondering whether you're encountering GC pauses that are causing some kind of increased slowdown.

Plotting your figures on a graph (after removing the initial outlier) looks like this:

**Lower is better**

<img width="584" alt="screen shot 2017-06-29 at 09 38 53" src="https://user-images.githubusercontent.com/719722/27678946-fd9f1b80-5cae-11e7-83af-45b1a536c844.png">

Whereas my own figures on a blank project (again, kit v1.10.1) look like this:

<img width="551" alt="screen shot 2017-06-29 at 09 41 35" src="https://user-images.githubusercontent.com/719722/27679007-3196107e-5caf-11e7-9245-3d04ded1f41f.png">

This is a small sample size, granted. I'm working on a project that will have a much more complex chain of components (potentially hundreds of them), so I'll re-run the benchmarks once that's in play.

**Comment by leebenson** [comment]
re: https://stackoverflow.com/questions/34728962/react-rendertostring-performance-and-caching-react-components

> In development, rendering ~1500 components takes ~200-300ms. By removing some components I was able to get ~1200 components to render in ~175-225ms.
>
> In production, renderToString on ~1500 components takes around ~50-200ms.

The general consensus is that `renderToString` _should_ be faster in production. Do a dump in your code to make sure you _are_ actually running in production:

```js
console.log('mode ->', process.env.NODE_ENV);
```

Are you using other libraries in your React components other than Apollo?

You might also try profiling your app using the [Node.js profiling guide](https://nodejs.org/en/docs/guides/simple-profiling/) or visually via a third-party app.

**Comment by balciseri** [comment]
Wow, you are so kind. I truly appreciate so much your help on this.

For the moment i'll leve my specs here:
Windows 10 Pro, Version 1703, SO Build 15063.413, X64 .
Ram 16 GB, i5-2500 @ 3.30Ghz
Node version: v8.0.0
Npm version: 5.0.4
ReactQl: v3.2.1
React : "^15.6.1",
Rreact-dom : "^15.6.1",

And this is the only code i added after running "reactql new" this morning, so no other libraries.
```
// Full React HTML render
console.log('mode ->', process.env.NODE_ENV);
const start = ms.now();
const html = ReactDOMServer.renderToString(components);
const end = ms.parse(ms.since(start));
const total = end.microseconds + (end.milliseconds * 1e3) + (end.seconds * 1e6);
console.log('total ->', total);
```
I can confirm i'm running in Prod mode when it's slower.

I'll update my node to match your version, and i'll do some more test
I will also look into the link you provided for profiling

**Comment by leebenson** [comment]
I wonder if it's something to do with Windows. Are you planning on hosting with Windows, or just developing locally?

Try building an image for Docker and running locally (https://reactql.org/docs/bundling/docker); I have a feeling you might get more consistent performance from Linux.

**Comment by balciseri** [comment]
I have not much experience with Docker.
I've uploaded my project with the complex data tree to my server which is running Ubuntu 16.04.1 LTS and node 8.1.3.

These are the result on linux.

Dev mode SSR query linux
getDataFromTree 687.154ms
renderToString 154.183ms
getDataFromTree 270.14ms
renderToString 112.698ms
getDataFromTree 291.082ms
renderToString 103.96ms
getDataFromTree 213.094ms
renderToString 85.916ms
getDataFromTree 223.412ms
renderToString 70.957ms
getDataFromTree 234.052ms
renderToString 80.039ms
getDataFromTree 203.536ms
renderToString 73.722ms
getDataFromTree 201.731ms
renderToString 74.057ms
getDataFromTree 148.426ms
renderToString 67.918ms
getDataFromTree 206.769ms
renderToString 69.631ms
getDataFromTree 274.677ms
renderToString 133.009ms
getDataFromTree 238.659ms
renderToString 82.969ms
getDataFromTree 289.843ms
renderToString 78.269ms
getDataFromTree 209.247ms
renderToString 77.672ms
getDataFromTree 183.97ms
renderToString 75.581ms
getDataFromTree 281.667ms
renderToString 78.488ms
getDataFromTree 203.532ms
renderToString 78.351ms
getDataFromTree 175.003ms
renderToString 75.869ms
getDataFromTree 209.86ms
renderToString 63.611ms
getDataFromTree 195.497ms
renderToString 82.904ms


Prod mode SSR query Linux
getDataFromTree 614.922ms
renderToString 154.721ms
getDataFromTree 323.278ms
renderToString 94.177ms
getDataFromTree 280.958ms
renderToString 75.865ms
getDataFromTree 204.78ms
renderToString 91.121ms
getDataFromTree 232.787ms
renderToString 84.01ms
getDataFromTree 191.031ms
renderToString 74.774ms
getDataFromTree 251.587ms
renderToString 69.564ms
getDataFromTree 225.188ms
renderToString 78.288ms
getDataFromTree 235.882ms
renderToString 87.051ms
getDataFromTree 186.104ms
renderToString 83.408ms
getDataFromTree 313.817ms
renderToString 132.573ms
getDataFromTree 277.965ms
renderToString 79.926ms
getDataFromTree 266.767ms
renderToString 67.088ms
getDataFromTree 206.499ms
renderToString 69.835ms
getDataFromTree 280.289ms
renderToString 69.496ms
getDataFromTree 218.952ms
renderToString 77.118ms
getDataFromTree 162.89ms
renderToString 64.621ms
getDataFromTree 259.277ms
renderToString 84.813ms
getDataFromTree 259.936ms
renderToString 59.467ms
getDataFromTree 175.486ms
renderToString 60.133ms

These are my result on windows:

Dev mode SSR query Windows
getDataFromTree 714.627ms
renderToString 162.605ms
getDataFromTree 271.402ms
renderToString 107.167ms
getDataFromTree 225.184ms
renderToString 91.654ms
getDataFromTree 169.644ms
renderToString 80.093ms
getDataFromTree 201.694ms
renderToString 81.667ms
getDataFromTree 178.412ms
renderToString 81.947ms
getDataFromTree 200.495ms
renderToString 92ms
getDataFromTree 204.448ms
renderToString 120.572ms
getDataFromTree 173.623ms
renderToString 76.771ms
getDataFromTree 225.592ms
renderToString 80ms
getDataFromTree 190.277ms
renderToString 99.025ms
getDataFromTree 181.874ms
renderToString 112.387ms
getDataFromTree 141.801ms
renderToString 63.398ms

Prod mode SSR query windows
getDataFromTree 655.533ms
renderToString 192.559ms
getDataFromTree 276.86ms
renderToString 226.536ms
getDataFromTree 172.774ms
renderToString 129.165ms
getDataFromTree 209.273ms
renderToString 144.75ms
getDataFromTree 190.636ms
renderToString 205.2ms
getDataFromTree 194.254ms
renderToString 136.911ms
getDataFromTree 212.961ms
renderToString 112.387ms
getDataFromTree 193.76ms
renderToString 154.484ms
getDataFromTree 171.173ms
renderToString 144.052ms
getDataFromTree 230.179ms
renderToString 156.367ms
getDataFromTree 182.737ms
renderToString 208.704ms
getDataFromTree 168.588ms
renderToString 127.103ms
getDataFromTree 134.658ms
renderToString 112.927ms

I didn't plotted the result, but looks like you were right.
On windows it's a lot highger the delay on Prod mode, but on Linux the results are quite similar

**Comment by leebenson** [comment]
Glad to see that it's working well on Linux.

I'm not sure what exactly the Windows issue is, but I'd always recommend running production code on *nix. In my local testing, Windows is anecdotally always extremely slow against comparable operations on Mac - even installing NPM generally takes 2-4x longer.

These results may change again (hopefully for the better!) when Facebook release their new React engine.
